### PR TITLE
feat(btcstaking): committing/timestamping public randomness via BIP-32

### LINKED ISSUE #591: BIP-32 based EOTS randomness derivation/timestamping
Proposal: https://babylon-chain.atlassian.net/wiki/spaces/BABYLON/pages/214630413/EOTS+randomness+derivation+protocol+from+BIP-32

Steps for implementation:

- Babylon side
  - [x] master secret/public randomness, DB schema, and message formats (#594 )
  - [x] Removing `MsgCommitPubRandList` and using master public randomness for verifying finality signatures (#595 )
  - [x] BTC timestamping master public randomness for finality providers (#598 )
  - [x] Docs and changelogs (#604 )
- Finality provider side
  - [x] master secret/public randomness, DB schema, and message formats in EOTS manager
  - [x] master secret/public randomness, DB schema, and message formats in finality provider
  - [x] Removing `MsgCommitPubRandList` and using master public randomness for verifying finality signatures
  - [x] BTC timestamping master public randomness for finality providers
  - [x] Docs and changelogs
- [ ] Bump vigilantes

### PR DESCRIPTION
**Pull request by SebastianElvis**
Fixes #591 

This PR provides the full implementation on deriving/timestamping EOTS public randomness via BIP-32. It consists of a series of small PRs (see below commit history).

NOTE: this should be merged via a merge commit

### CODE REVIEW

File: x/checkpointing/keeper/keeper.go [hunk 1]
```diff
@@ -395,3 +399,29 @@ func (k Keeper) GetTotalVotingPower(ctx context.Context, epochNumber uint64) int
...
+	if err != nil {
+		panic(err)
+	}
+	if !has {
+		return 0, types.ErrFinalizedEpochNotFound
+	}
+	epochNumberBytes, err := store.Get(types.LastFinalizedEpochKey)
+	if err != nil {
+		panic(err)
+	}
```
Comment (gitferry) [inline x/checkpointing/keeper/keeper.go:416]: Seems that `store.Get` is sufficient for our purpose. We can do nil check over `epochNumberBytes` to return `types.ErrFinalizedEpochNotFound`. Also, I think maybe we can panic on `types.ErrFinalizedEpochNotFound` if we decide to set epoch `0` finalized

File: x/btcstaking/client/cli/tx.go [hunk 2]
```diff
@@ -52,8 +52,8 @@ func GetTxCmd() *cobra.Command {
 
 func NewCreateFinalityProvicerCmd() *cobra.Command {
```
Comment (gitferry) [inline x/btcstaking/client/cli/tx.go:53]:
```suggestion
func NewCreateFinalityProviderCmd() *cobra.Command {
```

File: x/finality/keeper/msg_server.go [hunk 3]
```diff
@@ -77,6 +77,15 @@ func (ms msgServer) AddFinalitySig(goCtx context.Context, req *types.MsgAddFinal
 		return nil, bstypes.ErrFpAlreadySlashed
 	}
 
+	// ensure the finality provider's registered epoch is already finalised by BTC timestamping
+	finalizedEpoch, err := ms.BTCStakingKeeper.GetLastFinalizedEpoch(ctx)
```
Comment (gitferry) [inline x/finality/keeper/msg_server.go:81]: what if the finality sig is sent but epoch `1` has not been finalized? Will this lead to panic?

**Review by gitferry — APPROVED** [review]
Very neat set of prs! Only question is about finalizing epoch `0`

**Comment by SebastianElvis** [comment]
> Very neat set of prs! Only questions are around finalizing epoch `0`

Great point on finalising epoch 0! I have addressed the comment by finalising epoch 0 at genesis at https://github.com/babylonchain/babylon/pull/606. Please feel free to take a look

File: x/checkpointing/genesis.go [hunk 4]
```diff
@@ -10,6 +11,8 @@ import (
 // state.
 func InitGenesis(ctx context.Context, k keeper.Keeper, genState types.GenesisState) {
 	k.SetGenBlsKeys(ctx, genState.GenesisKeys)
+	// set epoch 0 to be finalised at genesis
+	k.SetLastFinalizedEpoch(ctx, 0)
```
[the remark this answers is not in the document]
Comment (SebastianElvis) [reply x/checkpointing/genesis.go:15]: Good point. Looks like the genesis import/export is not supported in the checkpointing module. This not only includes the last finalsied epoch, but also other storage, e.g., checkpoints. Left a TODO at https://github.com/babylonchain/babylon/pull/609 for now and need to address it in a separate PR

**Pull request merged**